// revolt.js/lib/cjs/collections/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserCollection = exports.SessionCollection = exports.ServerMemberCollection = exports.ServerCollection = exports.MessageCollection = exports.EmojiCollection = exports.ChannelWebhookCollection = exports.ChannelUnreadCollection = exports.ChannelCollection = exports.BotCollection = void 0;
__exportStar(require("./Collection"), exports);
var BotCollection_1 = require("./BotCollection");
Object.defineProperty(exports, "BotCollection", { enumerable: true, get: function () { return BotCollection_1.BotCollection; } });
var ChannelCollection_1 = require("./ChannelCollection");
Object.defineProperty(exports, "ChannelCollection", { enumerable: true, get: function () { return ChannelCollection_1.ChannelCollection; } });
var ChannelUnreadCollection_1 = require("./ChannelUnreadCollection");
Object.defineProperty(exports, "ChannelUnreadCollection", { enumerable: true, get: function () { return ChannelUnreadCollection_1.ChannelUnreadCollection; } });
var ChannelWebhookCollection_1 = require("./ChannelWebhookCollection");
Object.defineProperty(exports, "ChannelWebhookCollection", { enumerable: true, get: function () { return ChannelWebhookCollection_1.ChannelWebhookCollection; } });
var EmojiCollection_1 = require("./EmojiCollection");
Object.defineProperty(exports, "EmojiCollection", { enumerable: true, get: function () { return EmojiCollection_1.EmojiCollection; } });
var MessageCollection_1 = require("./MessageCollection");
Object.defineProperty(exports, "MessageCollection", { enumerable: true, get: function () { return MessageCollection_1.MessageCollection; } });
var ServerCollection_1 = require("./ServerCollection");
Object.defineProperty(exports, "ServerCollection", { enumerable: true, get: function () { return ServerCollection_1.ServerCollection; } });
var ServerMemberCollection_1 = require("./ServerMemberCollection");
Object.defineProperty(exports, "ServerMemberCollection", { enumerable: true, get: function () { return ServerMemberCollection_1.ServerMemberCollection; } });
var SessionCollection_1 = require("./SessionCollection");
Object.defineProperty(exports, "SessionCollection", { enumerable: true, get: function () { return SessionCollection_1.SessionCollection; } });
var UserCollection_1 = require("./UserCollection");
Object.defineProperty(exports, "UserCollection", { enumerable: true, get: function () { return UserCollection_1.UserCollection; } });
