// revolt.js/lib/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BotFlags = exports.UserFlags = exports.UserBadges = exports.ServerFlags = exports.ConnectionState = exports.EventClient = exports.Client = exports.API = void 0;
__exportStar(require("./classes"), exports);
__exportStar(require("./lib/regex"), exports);
__exportStar(require("./collections"), exports);
exports.API = __importStar(require("revolt-api"));
var Client_1 = require("./Client");
Object.defineProperty(exports, "Client", { enumerable: true, get: function () { return Client_1.Client; } });
var events_1 = require("./events");
Object.defineProperty(exports, "EventClient", { enumerable: true, get: function () { return events_1.EventClient; } });
Object.defineProperty(exports, "ConnectionState", { enumerable: true, get: function () { return events_1.ConnectionState; } });
var hydration_1 = require("./hydration");
Object.defineProperty(exports, "ServerFlags", { enumerable: true, get: function () { return hydration_1.ServerFlags; } });
Object.defineProperty(exports, "UserBadges", { enumerable: true, get: function () { return hydration_1.UserBadges; } });
Object.defineProperty(exports, "UserFlags", { enumerable: true, get: function () { return hydration_1.UserFlags; } });
Object.defineProperty(exports, "BotFlags", { enumerable: true, get: function () { return hydration_1.BotFlags; } });
