// revolt.js/lib/cjs/classes/ChannelWebhook.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _ChannelWebhook_collection;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChannelWebhook = void 0;
const hydration_1 = require("../hydration");
/**
 * Channel Webhook Class
 */
class ChannelWebhook {
    /**
     * Construct Channel Webhook
     * @param collection Collection
     * @param id Webhook
     */
    constructor(collection, id) {
        _ChannelWebhook_collection.set(this, void 0);
        __classPrivateFieldSet(this, _ChannelWebhook_collection, collection, "f");
        this.id = id;
    }
    /**
     * Webhook name
     */
    get name() {
        return __classPrivateFieldGet(this, _ChannelWebhook_collection, "f").getUnderlyingObject(this.id).name;
    }
    /**
     * Webhook avatar
     */
    get avatar() {
        return __classPrivateFieldGet(this, _ChannelWebhook_collection, "f").getUnderlyingObject(this.id).avatar;
    }
    /**
     * Webhook avatar URL
     */
    get avatarURL() {
        var _a;
        return (_a = __classPrivateFieldGet(this, _ChannelWebhook_collection, "f")
            .getUnderlyingObject(this.id)
            .avatar) === null || _a === void 0 ? void 0 : _a.createFileURL({ max_side: 256 });
    }
    /**
     * Channel ID this webhook belongs to
     */
    get channelId() {
        return __classPrivateFieldGet(this, _ChannelWebhook_collection, "f").getUnderlyingObject(this.id).channelId;
    }
    /**
     * Channel this webhook belongs to
     */
    get channel() {
        return __classPrivateFieldGet(this, _ChannelWebhook_collection, "f").client.channels.get(__classPrivateFieldGet(this, _ChannelWebhook_collection, "f").getUnderlyingObject(this.id).channelId);
    }
    /**
     * Secret token for sending messages to this webhook
     */
    get token() {
        return __classPrivateFieldGet(this, _ChannelWebhook_collection, "f").getUnderlyingObject(this.id).token;
    }
    /**
     * Edit this webhook
     */
    edit(data) {
        return __awaiter(this, void 0, void 0, function* () {
            const webhook = yield __classPrivateFieldGet(this, _ChannelWebhook_collection, "f").client.api.patch(`/webhooks/${this.id}/${this.token}`, data);
            __classPrivateFieldGet(this, _ChannelWebhook_collection, "f").updateUnderlyingObject(this.id, (0, hydration_1.hydrate)("channelWebhook", webhook, __classPrivateFieldGet(this, _ChannelWebhook_collection, "f").client));
        });
    }
    /**
     * Delete this webhook
     */
    delete() {
        return __awaiter(this, void 0, void 0, function* () {
            yield __classPrivateFieldGet(this, _ChannelWebhook_collection, "f").client.api.delete(`/webhooks/${this.id}/${this.token}`);
            __classPrivateFieldGet(this, _ChannelWebhook_collection, "f").delete(this.id);
        });
    }
}
exports.ChannelWebhook = ChannelWebhook;
_ChannelWebhook_collection = new WeakMap();
