// revolt.js/lib/cjs/classes/Message.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Message_collection, _MessageWebhook_client;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageWebhook = exports.Message = void 0;
const ulid_1 = require("ulid");
const __1 = require("..");
/**
 * Message Class
 */
class Message {
    /**
     * Construct Message
     * @param collection Collection
     * @param id Message Id
     */
    constructor(collection, id) {
        _Message_collection.set(this, void 0);
        __classPrivateFieldSet(this, _Message_collection, collection, "f");
        this.id = id;
    }
    /**
     * Time when this message was posted
     */
    get createdAt() {
        return new Date((0, ulid_1.decodeTime)(this.id));
    }
    /**
     * Absolute pathname to this message in the client
     */
    get path() {
        var _a;
        return `${(_a = this.channel) === null || _a === void 0 ? void 0 : _a.path}/${this.id}`;
    }
    /**
     * URL to this message
     */
    get url() {
        var _a;
        return ((_a = __classPrivateFieldGet(this, _Message_collection, "f").client.configuration) === null || _a === void 0 ? void 0 : _a.app) + this.path;
    }
    /**
     * Nonce value
     */
    get nonce() {
        return __classPrivateFieldGet(this, _Message_collection, "f").getUnderlyingObject(this.id).nonce;
    }
    /**
     * Id of channel this message was sent in
     */
    get channelId() {
        return __classPrivateFieldGet(this, _Message_collection, "f").getUnderlyingObject(this.id).channelId;
    }
    /**
     * Channel this message was sent in
     */
    get channel() {
        return __classPrivateFieldGet(this, _Message_collection, "f").client.channels.get(__classPrivateFieldGet(this, _Message_collection, "f").getUnderlyingObject(this.id).channelId);
    }
    /**
     * Server this message was sent in
     */
    get server() {
        var _a;
        return (_a = this.channel) === null || _a === void 0 ? void 0 : _a.server;
    }
    /**
     * Member this message was sent by
     */
    get member() {
        var _a;
        return __classPrivateFieldGet(this, _Message_collection, "f").client.serverMembers.getByKey({
            server: (_a = this.channel) === null || _a === void 0 ? void 0 : _a.serverId,
            user: this.authorId,
        });
    }
    /**
     * Id of user or webhook this message was sent by
     */
    get authorId() {
        return __classPrivateFieldGet(this, _Message_collection, "f").getUnderlyingObject(this.id).authorId;
    }
    /**
     * User this message was sent by
     */
    get author() {
        return __classPrivateFieldGet(this, _Message_collection, "f").client.users.get(__classPrivateFieldGet(this, _Message_collection, "f").getUnderlyingObject(this.id).authorId);
    }
    /**
     * Webhook information for this message
     */
    get webhook() {
        return __classPrivateFieldGet(this, _Message_collection, "f").getUnderlyingObject(this.id).webhook;
    }
    /**
     * Content
     */
    get content() {
        return __classPrivateFieldGet(this, _Message_collection, "f").getUnderlyingObject(this.id).content;
    }
    /**
     * System message content
     */
    get systemMessage() {
        return __classPrivateFieldGet(this, _Message_collection, "f").getUnderlyingObject(this.id).systemMessage;
    }
    /**
     * Attachments
     */
    get attachments() {
        return __classPrivateFieldGet(this, _Message_collection, "f").getUnderlyingObject(this.id).attachments;
    }
    /**
     * Time at which this message was edited
     */
    get editedAt() {
        return __classPrivateFieldGet(this, _Message_collection, "f").getUnderlyingObject(this.id).editedAt;
    }
    /**
     * Embeds
     */
    get embeds() {
        return __classPrivateFieldGet(this, _Message_collection, "f").getUnderlyingObject(this.id).embeds;
    }
    /**
     * IDs of users this message mentions
     */
    get mentionIds() {
        return __classPrivateFieldGet(this, _Message_collection, "f").getUnderlyingObject(this.id).mentionIds;
    }
    /**
     * IDs of messages this message replies to
     */
    get replyIds() {
        return __classPrivateFieldGet(this, _Message_collection, "f").getUnderlyingObject(this.id).replyIds;
    }
    /**
     * Reactions
     */
    get reactions() {
        return __classPrivateFieldGet(this, _Message_collection, "f").getUnderlyingObject(this.id).reactions;
    }
    /**
     * Interactions
     */
    get interactions() {
        return __classPrivateFieldGet(this, _Message_collection, "f").getUnderlyingObject(this.id).interactions;
    }
    /**
     * Masquerade
     */
    get masquerade() {
        return __classPrivateFieldGet(this, _Message_collection, "f").getUnderlyingObject(this.id).masquerade;
    }
    /**
     * Get the username for this message
     */
    get username() {
        var _a, _b, _c, _d, _e;
        const webhook = this.webhook;
        return ((_b = (_a = this.masquerade) === null || _a === void 0 ? void 0 : _a.name) !== null && _b !== void 0 ? _b : (webhook ? webhook.name : (_d = (_c = this.member) === null || _c === void 0 ? void 0 : _c.nickname) !== null && _d !== void 0 ? _d : (_e = this.author) === null || _e === void 0 ? void 0 : _e.username));
    }
    /**
     * Get the role colour for this message
     */
    get roleColour() {
        var _a, _b, _c;
        return (_b = (_a = this.masquerade) === null || _a === void 0 ? void 0 : _a.colour) !== null && _b !== void 0 ? _b : (_c = this.member) === null || _c === void 0 ? void 0 : _c.roleColour;
    }
    /**
     * Get the avatar URL for this message
     */
    get avatarURL() {
        var _a, _b, _c, _d;
        const webhook = this.webhook;
        return ((_a = this.masqueradeAvatarURL) !== null && _a !== void 0 ? _a : (webhook
            ? webhook.avatarURL
            : (_c = (_b = this.member) === null || _b === void 0 ? void 0 : _b.avatarURL) !== null && _c !== void 0 ? _c : (_d = this.author) === null || _d === void 0 ? void 0 : _d.avatarURL));
    }
    /**
     * Get the animated avatar URL for this message
     */
    get animatedAvatarURL() {
        var _a, _b, _c;
        const webhook = this.webhook;
        return ((_a = this.masqueradeAvatarURL) !== null && _a !== void 0 ? _a : (webhook
            ? webhook.avatarURL
            : this.member
                ? (_b = this.member) === null || _b === void 0 ? void 0 : _b.animatedAvatarURL
                : (_c = this.author) === null || _c === void 0 ? void 0 : _c.animatedAvatarURL));
    }
    /**
     * Avatar URL from the masquerade
     */
    get masqueradeAvatarURL() {
        var _a;
        const avatar = (_a = this.masquerade) === null || _a === void 0 ? void 0 : _a.avatar;
        return avatar ? __classPrivateFieldGet(this, _Message_collection, "f").client.proxyFile(avatar) : undefined;
    }
    /**
     * Edit a message
     * @param data Message edit route data
     */
    edit(data) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield __classPrivateFieldGet(this, _Message_collection, "f").client.api.patch(`/channels/${this.channelId}/messages/${this.id}`, data);
        });
    }
    /**
     * Delete a message
     */
    delete() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield __classPrivateFieldGet(this, _Message_collection, "f").client.api.delete(`/channels/${this.channelId}/messages/${this.id}`);
        });
    }
    /**
     * Acknowledge this message as read
     */
    ack() {
        var _a;
        (_a = this.channel) === null || _a === void 0 ? void 0 : _a.ack(this);
    }
    /**
     * Reply to Message
     */
    reply(data, mention = true) {
        var _a;
        const obj = typeof data === "string" ? { content: data } : data;
        return (_a = this.channel) === null || _a === void 0 ? void 0 : _a.sendMessage(Object.assign(Object.assign({}, obj), { replies: [{ id: this.id, mention }] }));
    }
    /**
     * Clear all reactions from this message
     */
    clearReactions() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield __classPrivateFieldGet(this, _Message_collection, "f").client.api.delete(`/channels/${this.channelId}/messages/${this.id}/reactions`);
        });
    }
    /**
     * React to a message
     * @param emoji Unicode or emoji ID
     */
    react(emoji) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield __classPrivateFieldGet(this, _Message_collection, "f").client.api.put(`/channels/${this.channelId}/messages/${this.id}/reactions/${emoji}`);
        });
    }
    /**
     * Un-react from a message
     * @param emoji Unicode or emoji ID
     */
    unreact(emoji) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield __classPrivateFieldGet(this, _Message_collection, "f").client.api.delete(`/channels/${this.channelId}/messages/${this.id}/reactions/${emoji}`);
        });
    }
}
exports.Message = Message;
_Message_collection = new WeakMap();
/**
 * Message Webhook Class
 */
class MessageWebhook {
    /**
     * Construct Message Webhook
     * @param client Client
     * @param webhook Webhook data
     */
    constructor(client, webhook, id) {
        _MessageWebhook_client.set(this, void 0);
        __classPrivateFieldSet(this, _MessageWebhook_client, client, "f");
        this.id = id;
        this.name = webhook.name;
        this.avatar = webhook.avatar
            ? new __1.File(client, {
                _id: webhook.avatar,
                tag: "avatars",
                metadata: {
                    type: "Image",
                    width: 256,
                    height: 256,
                },
            })
            : undefined;
    }
    /**
     * Get the avatar URL for this message webhook
     */
    get avatarURL() {
        var _a, _b;
        return ((_b = (_a = this.avatar) === null || _a === void 0 ? void 0 : _a.createFileURL({ max_side: 256 })) !== null && _b !== void 0 ? _b : `${__classPrivateFieldGet(this, _MessageWebhook_client, "f").options.baseURL}/users/${this.id}/default_avatar`);
    }
}
exports.MessageWebhook = MessageWebhook;
_MessageWebhook_client = new WeakMap();
