// revolt.js/lib/cjs/classes/ServerBan.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerBan = void 0;
/**
 * Server Ban
 */
class ServerBan {
    /**
     * Construct Server Ban
     * @param client Client
     * @param data Data
     */
    constructor(client, data) {
        this.client = client;
        this.id = data._id;
        this.reason = data.reason;
    }
    /**
     * User
     */
    get user() {
        return this.client.users.get(this.id.user);
    }
    /**
     * Server
     */
    get server() {
        return this.client.servers.get(this.id.server);
    }
    /**
     * Remove this server ban
     */
    pardon() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.client.api.delete(`/servers/${this.id.server}/bans/${this.id.user}`);
        });
    }
}
exports.ServerBan = ServerBan;
