// revolt.js/lib/cjs/classes/SystemMessage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChannelOwnershipChangeSystemMessage = exports.ChannelRenamedSystemMessage = exports.ChannelEditSystemMessage = exports.UserModeratedSystemMessage = exports.UserSystemMessage = exports.TextSystemMessage = exports.SystemMessage = void 0;
/**
 * System Message
 */
class SystemMessage {
    /**
     * Construct System Message
     * @param client Client
     * @param type Type
     */
    constructor(client, type) {
        this.client = client;
        this.type = type;
    }
    /**
     * Create an System Message from an API System Message
     * @param client Client
     * @param embed Data
     * @returns System Message
     */
    static from(client, message) {
        switch (message.type) {
            case "text":
                return new TextSystemMessage(client, message);
            case "user_added":
            case "user_remove":
                return new UserModeratedSystemMessage(client, message);
            case "user_joined":
            case "user_left":
            case "user_kicked":
            case "user_banned":
                return new UserSystemMessage(client, message);
            case "channel_renamed":
                return new ChannelRenamedSystemMessage(client, message);
            case "channel_description_changed":
            case "channel_icon_changed":
                return new ChannelEditSystemMessage(client, message);
            case "channel_ownership_changed":
                return new ChannelOwnershipChangeSystemMessage(client, message);
            default:
                return new TextSystemMessage(client, {
                    type: "text",
                    content: `${message.type} is not supported.`,
                });
        }
    }
}
exports.SystemMessage = SystemMessage;
/**
 * Text System Message
 */
class TextSystemMessage extends SystemMessage {
    /**
     * Construct System Message
     * @param client Client
     * @param systemMessage System Message
     */
    constructor(client, systemMessage) {
        super(client, systemMessage.type);
        this.content = systemMessage.content;
    }
}
exports.TextSystemMessage = TextSystemMessage;
/**
 * User System Message
 */
class UserSystemMessage extends SystemMessage {
    /**
     * Construct System Message
     * @param client Client
     * @param systemMessage System Message
     */
    constructor(client, systemMessage) {
        super(client, systemMessage.type);
        this.userId = systemMessage.id;
    }
    /**
     * User this message concerns
     */
    get user() {
        return this.client.users.get(this.userId);
    }
}
exports.UserSystemMessage = UserSystemMessage;
/**
 * User Moderated System Message
 */
class UserModeratedSystemMessage extends UserSystemMessage {
    /**
     * Construct System Message
     * @param client Client
     * @param systemMessage System Message
     */
    constructor(client, systemMessage) {
        super(client, systemMessage);
        this.byId = systemMessage.by;
    }
    /**
     * User this action was performed by
     */
    get by() {
        console.info("deez!", this.byId);
        return this.client.users.get(this.byId);
    }
}
exports.UserModeratedSystemMessage = UserModeratedSystemMessage;
/**
 * Channel Edit System Message
 */
class ChannelEditSystemMessage extends SystemMessage {
    /**
     * Construct System Message
     * @param client Client
     * @param systemMessage System Message
     */
    constructor(client, systemMessage) {
        super(client, systemMessage.type);
        this.byId = systemMessage.by;
    }
    /**
     * User this action was performed by
     */
    get by() {
        return this.client.users.get(this.byId);
    }
}
exports.ChannelEditSystemMessage = ChannelEditSystemMessage;
/**
 * Channel Renamed System Message
 */
class ChannelRenamedSystemMessage extends ChannelEditSystemMessage {
    /**
     * Construct System Message
     * @param client Client
     * @param systemMessage System Message
     */
    constructor(client, systemMessage) {
        super(client, systemMessage);
        this.name = systemMessage.name;
    }
}
exports.ChannelRenamedSystemMessage = ChannelRenamedSystemMessage;
/**
 * Channel Ownership Change System Message
 */
class ChannelOwnershipChangeSystemMessage extends SystemMessage {
    /**
     * Construct System Message
     * @param client Client
     * @param systemMessage System Message
     */
    constructor(client, systemMessage) {
        super(client, systemMessage.type);
        this.fromId = systemMessage.from;
        this.toId = systemMessage.to;
    }
    /**
     * User giving away channel ownership
     */
    get from() {
        return this.client.users.get(this.fromId);
    }
    /**
     * User receiving channel ownership
     */
    get to() {
        return this.client.users.get(this.toId);
    }
}
exports.ChannelOwnershipChangeSystemMessage = ChannelOwnershipChangeSystemMessage;
