// revolt.js/lib/cjs/hydration/channel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.channelHydration = void 0;
const set_1 = require("@solid-primitives/set");
const __1 = require("..");
exports.channelHydration = {
    keyMapping: {
        _id: "id",
        channel_type: "channelType",
        recipients: "recipientIds",
        user: "userId",
        owner: "ownerId",
        server: "serverId",
        default_permissions: "defaultPermissions",
        role_permissions: "rolePermissions",
        last_message_id: "lastMessageId",
    },
    functions: {
        id: (channel) => channel._id,
        channelType: (channel) => channel.channel_type,
        name: (channel) => channel.name,
        description: (channel) => channel.description,
        icon: (channel, ctx) => new __1.File(ctx, channel.icon),
        active: (channel) => channel.active || false,
        typingIds: () => new set_1.ReactiveSet(),
        recipientIds: (channel) => new set_1.ReactiveSet(channel.recipients),
        userId: (channel) => channel.user,
        ownerId: (channel) => channel.owner,
        serverId: (channel) => channel.server,
        permissions: (channel) => channel.permissions,
        defaultPermissions: (channel) => channel.default_permissions,
        rolePermissions: (channel) => channel.role_permissions,
        nsfw: (channel) => channel.nsfw || false,
        lastMessageId: (channel) => channel.last_message_id,
    },
    initialHydration: () => ({
        typingIds: new set_1.ReactiveSet(),
        recipientIds: new set_1.ReactiveSet(),
    }),
};
