// revolt.js/lib/cjs/hydration/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hydrate = exports.UserFlags = exports.UserBadges = exports.ServerFlags = exports.BotFlags = void 0;
const bot_1 = require("./bot");
const channel_1 = require("./channel");
const channelUnread_1 = require("./channelUnread");
const channelWebhook_1 = require("./channelWebhook");
const emoji_1 = require("./emoji");
const message_1 = require("./message");
const server_1 = require("./server");
const serverMember_1 = require("./serverMember");
const session_1 = require("./session");
const user_1 = require("./user");
var bot_2 = require("./bot");
Object.defineProperty(exports, "BotFlags", { enumerable: true, get: function () { return bot_2.BotFlags; } });
var server_2 = require("./server");
Object.defineProperty(exports, "ServerFlags", { enumerable: true, get: function () { return server_2.ServerFlags; } });
var user_2 = require("./user");
Object.defineProperty(exports, "UserBadges", { enumerable: true, get: function () { return user_2.UserBadges; } });
Object.defineProperty(exports, "UserFlags", { enumerable: true, get: function () { return user_2.UserFlags; } });
/**
 * Hydrate some data
 * @param hydration Hydration data
 * @param input Input data
 * @returns Output data
 */
function hydrateInternal(hydration, input, context) {
    return Object.keys(input).reduce((acc, key) => {
        var _a;
        let targetKey, value;
        try {
            targetKey = (_a = hydration.keyMapping[key]) !== null && _a !== void 0 ? _a : key;
            value = hydration.functions[targetKey](input, context);
        }
        catch (err) {
            if (key === "type")
                return acc;
            console.debug(`Skipping key ${String(key)} during hydration!`);
            return acc;
        }
        return Object.assign(Object.assign({}, acc), { [targetKey]: value });
    }, {});
}
const hydrators = {
    bot: bot_1.botHydration,
    channel: channel_1.channelHydration,
    channelUnread: channelUnread_1.channelUnreadHydration,
    channelWebhook: channelWebhook_1.channelWebhookHydration,
    emoji: emoji_1.emojiHydration,
    message: message_1.messageHydration,
    server: server_1.serverHydration,
    serverMember: serverMember_1.serverMemberHydration,
    session: session_1.sessionHydration,
    user: user_1.userHydration,
};
/**
 * Hydrate some input with a given type
 * @param type Type
 * @param input Input Object
 * @param initial Whether this is the initial hydration
 * @returns Hydrated Object
 */
function hydrate(type, input, context, initial) {
    return hydrateInternal(hydrators[type], initial ? Object.assign(Object.assign({}, hydrators[type].initialHydration()), input) : input, context);
}
exports.hydrate = hydrate;
