// revolt.js/lib/cjs/hydration/message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.messageHydration = void 0;
const map_1 = require("@solid-primitives/map");
const set_1 = require("@solid-primitives/set");
const __1 = require("..");
exports.messageHydration = {
    keyMapping: {
        _id: "id",
        channel: "channelId",
        author: "authorId",
        system: "systemMessage",
        edited: "editedAt",
        mentions: "mentionIds",
        replies: "replyIds",
    },
    functions: {
        id: (message) => message._id,
        nonce: (message) => message.nonce,
        channelId: (message) => message.channel,
        authorId: (message) => message.author,
        webhook: (message, ctx) => message.webhook
            ? new __1.MessageWebhook(ctx, message.webhook, message.author)
            : undefined,
        content: (message) => message.content,
        systemMessage: (message, ctx) => __1.SystemMessage.from(ctx, message.system),
        attachments: (message, ctx) => message.attachments.map((file) => new __1.File(ctx, file)),
        editedAt: (message) => new Date(message.edited),
        embeds: (message, ctx) => message.embeds.map((embed) => __1.MessageEmbed.from(ctx, embed)),
        mentionIds: (message) => message.mentions,
        replyIds: (message) => message.replies,
        reactions: (message) => {
            const map = new map_1.ReactiveMap();
            if (message.reactions) {
                for (const reaction of Object.keys(message.reactions)) {
                    map.set(reaction, new set_1.ReactiveSet(message.reactions[reaction]));
                }
            }
            return map;
        },
        interactions: (message) => message.interactions,
        masquerade: (message) => message.masquerade,
    },
    initialHydration: () => ({
        reactions: new map_1.ReactiveMap(),
    }),
};
