// revolt.js/lib/cjs/hydration/server.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerFlags = exports.serverHydration = void 0;
const map_1 = require("@solid-primitives/map");
const set_1 = require("@solid-primitives/set");
const __1 = require("..");
exports.serverHydration = {
    keyMapping: {
        _id: "id",
        owner: "ownerId",
        channels: "channelIds",
        system_messages: "systemMessages",
        default_permissions: "defaultPermissions",
    },
    functions: {
        id: (server) => server._id,
        ownerId: (server) => server.owner,
        name: (server) => server.name,
        description: (server) => server.description,
        channelIds: (server) => new set_1.ReactiveSet(server.channels),
        categories: (server) => { var _a; return (_a = server.categories) !== null && _a !== void 0 ? _a : []; },
        systemMessages: (server) => { var _a; return (_a = server.system_messages) !== null && _a !== void 0 ? _a : {}; },
        roles: (server) => new map_1.ReactiveMap(Object.keys(server.roles).map((id) => [id, server.roles[id]])),
        defaultPermissions: (server) => server.default_permissions,
        icon: (server, ctx) => new __1.File(ctx, server.icon),
        banner: (server, ctx) => new __1.File(ctx, server.banner),
        flags: (server) => server.flags,
        analytics: (server) => server.analytics || false,
        discoverable: (server) => server.discoverable || false,
        nsfw: (server) => server.nsfw || false,
    },
    initialHydration: () => ({
        channelIds: new set_1.ReactiveSet(),
        roles: new map_1.ReactiveMap(),
    }),
};
/**
 * Flags attributed to servers
 */
var ServerFlags;
(function (ServerFlags) {
    ServerFlags[ServerFlags["Official"] = 1] = "Official";
    ServerFlags[ServerFlags["Verified"] = 2] = "Verified";
})(ServerFlags = exports.ServerFlags || (exports.ServerFlags = {}));
