// revolt.js/lib/hydration/channel copy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.channelHydration = void 0;
/**
 * Mapping from API Channel to hydrated Channel
 */
exports.channelHydration = {
    keyMapping: {
        _id: "id",
        channel_type: "channelType",
        recipients: "recipientIds",
        owner: "ownerId",
        server: "serverId",
        default_permissions: "defaultPermissions",
        role_permissions: "rolePermissions",
        last_message_id: "lastMessageId",
    },
    functions: {
        id: (channel) => channel._id,
        channelType: (channel) => channel.channel_type,
        name: (channel) => channel.name,
        description: (channel) => channel.description,
        icon: (channel) => channel.icon,
        active: (channel) => channel.active || false,
        recipientIds: (channel) => { var _a; return (_a = channel.recipients) !== null && _a !== void 0 ? _a : []; },
        ownerId: (channel) => channel.owner,
        serverId: (channel) => channel.server,
        permissions: (channel) => channel.permissions,
        defaultPermissions: (channel) => channel.default_permissions,
        rolePermissions: (channel) => { var _a; return (_a = channel.role_permissions) !== null && _a !== void 0 ? _a : {}; },
        nsfw: (channel) => channel.nsfw || false,
        lastMessageId: (channel) => channel.last_message_id,
    },
};
