// revolt.js/lib/hydration/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hydrate = void 0;
const channel_1 = require("./channel");
const emoji_1 = require("./emoji");
const message_1 = require("./message");
const server_1 = require("./server");
const serverMember_1 = require("./serverMember");
const user_1 = require("./user");
/**
 * Hydrate some data
 * @param hydration Hydration data
 * @param input Input data
 * @returns Output data
 */
function hydrateInternal(hydration, input) {
    return Object.keys(input).reduce((acc, key) => {
        var _a;
        let targetKey, value;
        try {
            targetKey = (_a = hydration.keyMapping[key]) !== null && _a !== void 0 ? _a : key;
            value = hydration.functions[targetKey](input);
        }
        catch (err) {
            console.debug(`Skipping key ${String(key)} during hydration!`);
            return acc;
        }
        return Object.assign(Object.assign({}, acc), { [targetKey]: value });
    }, {});
}
const hydrators = {
    channel: channel_1.channelHydration,
    emoji: emoji_1.emojiHydration,
    message: message_1.messageHydration,
    server: server_1.serverHydration,
    serverMember: serverMember_1.serverMemberHydration,
    user: user_1.userHydration,
};
/**
 * Hydrate some input with a given type
 * @param type Type
 * @param input Input Object
 * @returns Hydrated Object
 */
function hydrate(type, input) {
    return hydrateInternal(hydrators[type], input);
}
exports.hydrate = hydrate;
