// revolt.js/lib/hydration/message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.messageHydration = void 0;
const map_1 = require("@solid-primitives/map");
const set_1 = require("@solid-primitives/set");
exports.messageHydration = {
    keyMapping: {
        _id: "id",
        channel: "channelId",
        author: "authorId",
        system: "systemMessage",
        edited: "editedAt",
        mentions: "mentionIds",
        replies: "replyIds",
    },
    functions: {
        id: (message) => message._id,
        nonce: (message) => message.nonce,
        channelId: (message) => message.channel,
        authorId: (message) => message.author,
        content: (message) => message.content,
        systemMessage: (message) => message.system,
        attachments: (message) => message.attachments,
        editedAt: (message) => new Date(message.edited),
        embeds: (message) => message.embeds,
        mentionIds: (message) => message.mentions,
        replyIds: (message) => message.replies,
        reactions: (message) => {
            const map = new map_1.ReactiveMap();
            if (message.reactions) {
                for (const reaction of Object.keys(message.reactions)) {
                    map.set(reaction, new set_1.ReactiveSet(message.reactions[reaction]));
                }
            }
            return map;
        },
        interactions: (message) => message.interactions,
        masquerade: (message) => message.masquerade,
    },
};
