// revolt.js/lib/hydration/user.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserFlags = exports.UserBadges = exports.userHydration = void 0;
exports.userHydration = {
    keyMapping: {
        _id: "id",
    },
    functions: {
        id: (user) => user._id,
        username: (user) => user.username,
        relationship: (user) => { var _a; return (_a = user.relationship) !== null && _a !== void 0 ? _a : "None"; },
        online: (user) => user.online || false,
        privileged: (user) => user.privileged || false,
        badges: (user) => { var _a; return (_a = user.badges) !== null && _a !== void 0 ? _a : 0; },
        flags: (user) => { var _a; return (_a = user.flags) !== null && _a !== void 0 ? _a : 0; },
        avatar: (user) => { var _a; return (_a = user.avatar) !== null && _a !== void 0 ? _a : undefined; },
        status: (user) => { var _a; return (_a = user.status) !== null && _a !== void 0 ? _a : undefined; },
        bot: (user) => { var _a; return (_a = user.bot) !== null && _a !== void 0 ? _a : undefined; },
    },
};
/**
 * Badges available to users
 */
var UserBadges;
(function (UserBadges) {
    UserBadges[UserBadges["Developer"] = 1] = "Developer";
    UserBadges[UserBadges["Translator"] = 2] = "Translator";
    UserBadges[UserBadges["Supporter"] = 4] = "Supporter";
    UserBadges[UserBadges["ResponsibleDisclosure"] = 8] = "ResponsibleDisclosure";
    UserBadges[UserBadges["Founder"] = 16] = "Founder";
    UserBadges[UserBadges["PlatformModeration"] = 32] = "PlatformModeration";
    UserBadges[UserBadges["ActiveSupporter"] = 64] = "ActiveSupporter";
    UserBadges[UserBadges["Paw"] = 128] = "Paw";
    UserBadges[UserBadges["EarlyAdopter"] = 256] = "EarlyAdopter";
    UserBadges[UserBadges["ReservedRelevantJokeBadge1"] = 512] = "ReservedRelevantJokeBadge1";
    UserBadges[UserBadges["ReservedRelevantJokeBadge2"] = 1024] = "ReservedRelevantJokeBadge2";
})(UserBadges = exports.UserBadges || (exports.UserBadges = {}));
/**
 * Flags attributed to users
 */
var UserFlags;
(function (UserFlags) {
    UserFlags[UserFlags["Suspended"] = 1] = "Suspended";
    UserFlags[UserFlags["Deleted"] = 2] = "Deleted";
    UserFlags[UserFlags["Banned"] = 4] = "Banned";
})(UserFlags = exports.UserFlags || (exports.UserFlags = {}));
